### ISSUE TITLE
Do not start queries if there is not enough memory

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#52254]
**Use case**

When memory of server is exhausted i.e. current memory is higher that max memory limit do not allow to start new queries.

**Describe the solution you'd like**

When we add queries to `ProcessList` we check limit on concurrent queries and wait for `queue_max_wait_ms` to get free slot and throw `TOO_MANY_SIMULTANEOUS_QUERIES` if there is no slot available after waiting.

We can add similar logic that checks global MemoryTracker (or even user Memory Tracker) before we start running query. If there is no memory we can throw `SERVER_MEMORY_OVERLOADED` or something similar, but not `MEMORY_LIMIT_EXCEEDED` as it is guaranteed that query was not running.

We can even add separate setting `min_free_memory_for_query` that checks that at least that amount of free memory exists when we start a query. It is not as good as https://github.com/ClickHouse/ClickHouse/issues/41887 because we will start multiple queries from a queue all at once when we have enough memory. But it is not worse than current implementation that starts queries without checking available memory at all.

https://github.com/ClickHouse/ClickHouse/blob/37054e7b47311b941ba2096a2816ba273a387489/src/Interpreters/ProcessList.cpp#L69

**Comment by WangTaoTheTonic** [comment]
Nice! This ability has mint of requirements! We can also considering query queue like this: https://docs.starrocks.io/en-us/3.0/administration/query_queues It is superset of just checking memory.

**Comment by rvasin** [comment]
The task is very interesting and useful. The main question: before starting a query may we know for sure how much memory will it take?  Or should we use some heuristics? 
@serxa maybe you may answer this question as author of #41887

**Comment by serxa** [comment]
We definitely cannot tell, but we can do estimates and try to improve them as we go down this road. But IMO the main idea is to introduce a couple of new concepts. Every query should estimate `Memory Demand`: how much memory it will require during execution (simple constant to begin with). Demands of concurrent queries are summed to a total and server will block arriving queries with some timeout if sum will exceed `Guaranteed Memory Limit`. I think we can even expose this Memory Demand as a query setting with `auto` value meaning "Dear ClickHouse, please estimate my query memory demand"

**Comment by Seb-SK** [comment]
Hi guys,

Just a slight up to say we would be very interested in this as well !

Use case : looker as front end, our own service with clickhouse as back-end.
90% of our dashboards use simple queries and we don't have any issues, but one is using a few complex queries that take quite a bit of time (~5s) and a lot of RAM (each query takes ~5 to 9Gb of memory, the dashboard has ~20 widgets).
Looker tries to load all the widgets on one page in parallel, which means we receive a "burst of queries" at the same time.
Some of them (actually a lot of them) will therefore fail with a "(MEMORY_LIMIT_EXCEEDED) error.

The problem is that we can't do much here. We cannot control Looker's behavior, we did try to put a "middleware" in our BI service to force only 2 or 3 queries to be processed at the same time, but then this will impact all the dashboards which have much simpler queries (we didn't use the max_concurrent_queries because that was also blocking our inserts, and max_concurrent_select_queries was triggering exceptions instead of queuing the queries).

The solution would really be for clickhouse to simply queue queries until enough memory is freed so that it can execute them (with a maximum timeout value).

**Comment by qoega** [comment]
>We can also consider more complex query queue

It is out of this issue scope. You can create a separate issue with RFC of queries queue.

>before starting a query may we know for sure how much memory will it take?

No, we will not have estimate. It is just a way to not start new queries when service is already overloaded and return it earlier than starting and returning MEMORY_LIMIT_EXCEEDED

>The solution would really be for clickhouse to simply queue queries until enough memory is freed so that it can execute them (with a maximum timeout value).

This issue is not about queries queue. The feature discussed here will return SERVER_MEMORY_OVERLOADED part of your queries burst.

**Comment by serxa** [comment]
@Seb-SK [workload scheduling](https://clickhouse.com/docs/operations/workload-scheduling) will eventually have a way to limit the number of concurrent queries (and queries waiting in the queue) for a specific workload. I'm currently working on the [CPU scheduling](https://github.com/ClickHouse/ClickHouse/pull/77595), and the next will be "Query Slot" scheduling. It should help to resolve your problem.
